thoth/sourcemanagement/sourcemanagement.py

```diff
@@ -41,33 +45,70 @@
 class SourceManagement:
     """Abstract source code management services like GitHub and GitLab."""
 
-    def __init__(self, service_type: ServiceType, service_url: str, token: str, slug: str):
+    def __init__(
+        self, service_type: ServiceType, service_url: str, slug: str, token: Optional[str], installation: bool = True,
+    ):
         """Initialize source code management tools abstraction.
 
         Note that we are using OGR for calls. OGR keeps URL to services in its global context per GitHub/GitLab.
         This is global context is initialized in the manager with a hope to fix this behavior for our needs.
         """
         self.service_type = service_type
+        self.service_url = service_url
         self.slug = slug
         self.service_url = service_url
         self.token = token
         self.namespace, self.repo = slug.rsplit("/", 1)
-
+        self.installation = installation
+        # token expires after 10 mins
+        self.token_expire_time = datetime.datetime.now() + datetime.timedelta(minutes=9, seconds=30)
+        self.github_auth_obj = None
+
+        if not installation and not token:
+            raise ValueError("Token nor Installation ID found during initialization.")
+        if installation:
+            self.github_auth_obj = GithubAuthentication(self.slug)
+            self.token = self.github_auth_obj.get_access_token()
+
+        # Initialize ogr service object
+        self._init_helper()
+
+    def _init_helper(self):
+        """Handle the initialization or reinitialization of OGR object."""
         if self.service_type == ServiceType.GITHUB:
-            if service_url:
-                self.service = GithubService(self.token, instance_url=service_url)
+            if self.service_url:
+                self.service = GithubService(self.token, instance_url=self.service_url)
             else:
                 self.service = GithubService(self.token)
             self.repository = self.service.get_project(repo=self.repo, namespace=self.namespace)
         elif self.service_type == ServiceType.GITLAB:
-            if service_url:
-                self.service = GitlabService(self.token, instance_url=service_url)
+            if self.service_url:
+                self.service = GitlabService(self.token, instance_url=self.service_url)
             else:
                 self.service = GitlabService(self.token)
             self.repository = self.service.get_project(repo=self.repo, namespace=self.namespace)
         else:
             raise NotImplementedError
 
+    class Token:
```

Comment:
Please move this nested class outside - nested classes/functions are hard to test and this introduces unnecessary complexity. You can also use just a function for decoration as you have only one staticmethod.

Reply:
Thanks, I removed the class.